tensorboard/webapp/metrics/views/card_renderer/histogram_card_container.ts

```diff
@@ -150,18 +157,43 @@ export class HistogramCardContainer implements CardRenderer, OnInit {
       })
     );
 
-    this.selectedTime$ = this.store.select(getMetricsSelectedTime).pipe(
-      combineLatestWith(this.data$),
-      map(([selectedTime, data]) => {
+    this.steps$ = this.data$.pipe(
+      map((data) => data.map((datum) => datum.step))
+    );
+
+    this.closestNonSelectedStep$ = combineLatest([
+      this.store.select(getMetricsSelectedTime),
+      this.steps$,
+    ]).pipe(
+      map(([selectedTime, steps]) => {
+        // Only calculates the closest step in single selection.
+        if (!selectedTime || !selectedTime.start || selectedTime.end !== null)
+          return null;
+
+        return getClosestNonSelectedStep(selectedTime.start.step, steps);
+      })
+    );
+
+    this.selectedTime$ = combineLatest([
+      this.store.select(getMetricsSelectedTime),
+      this.steps$,
+      this.closestNonSelectedStep$,
+    ]).pipe(
+      map(([selectedTime, steps, closestStep]) => {
         if (!selectedTime) return null;
 
         let minStep = Infinity;
         let maxStep = -Infinity;
-        for (const datum of data) {
-          minStep = Math.min(datum.step, minStep);
-          maxStep = Math.max(datum.step, maxStep);
+        for (const step of steps) {
+          minStep = Math.min(step, minStep);
+          maxStep = Math.max(step, maxStep);
         }
-        return maybeClipSelectedTime(selectedTime, minStep, maxStep);
+        const viewSelectedTime = maybeClipSelectedTime(
+          selectedTime,
+          minStep,
+          maxStep
+        );
+        return maybeSetClosestStartStep(viewSelectedTime, closestStep);
```

Comment:
This can be greatly simplified.

`if (viewSelectedTime.endStep === null) {
          viewSelectedTime.startStep = getClosestStep(
            viewSelectedTime.startStep,
            steps
          );
        }`

getClosestStep is like getClosestNonSelectedStep except if the selected step exists it just return it.

`export function getClosestStep(selectedStep: number, steps: number[]): number {
  let minDistance = Infinity;
  let closestStep = null;
  for (const step of steps) {
    const distance = Math.abs(selectedStep - step);
    if (distance < minDistance) {
      minDistance = distance;
      closestStep = step;
    }
  }

  return closestStep || selectedStep;
}`

Then you can get rid of the closestNonSelectedStep$ variable entirely and both of those utility functions.

Let me know what you think.

Reply:
1.I considered similar idea of your "getClosestStep" but I didn't go with it. It could be even simpler that `closestStep` could be the same as `selectedStep` (distance is 0)

The reason why I need "closestNonSelectedStep" instead of "closestStep" is I need to know if the step is "adjusted" at all. (To show the indicator) 

concept like: "isClosestStepHighLighted$ =true when closestNonSelectedStep$ !== null"

2. the reason why I prefer not put condition in your first code block is that it is less readable. What does it mean that "endStep===null"? (believe me, Brian asked the same question) Why do I set startStep when the condition is true? then we ended up adding more comments while I think the function name and docstring could be a better place to explain.

Reply:
1. Oh yap, that sounds a good idea. Thanks!
2. TBH I still prefer wrapping that part in a method to keep this block shorter and readable. I also want to update "maybeSetClosestStartStep" a bit to avoid in-place modification.

Reply:
2. I personally think that it is more readable to say if not range selection set start step to closest step.  However, that sounds like personal preference so I am ok with you doing it your way. Carry on.

Reply:
2. Thanks. I also want to avoid some confusion, like, if "the closest step" is start step itself, technically we don't need to anything. Assigning it to startStep no matter what looks confusing. (Although practically the value does not get changed). So wrapping it in a function with a "maybe" statement makes more sense to me.